### PR TITLE
RR-236 - Refactored to use a DTO when creating an action plan and goal 

### PR DESCRIPTION
**Pull request by nathanrussell-moj-digital**
This PR refactors the way we create Action Plans and Goals to use a DTO instead of directly creating a Goal domain object which may have fields we cannot set at this point.

The follow up to this PR will be a very similar one that makes the same type of changes to the Update Goal process (so that we use a DTO). I decided not to do them both in the same PR as this PR is big enough already! 😬 

Once both creating and updating have been changed to use DTOs, then we can update the DTO, domain classes and entities and associated mappers, to map the prisonId from the request objects so that we record the prison ID on create and update

### CODE REVIEW

File: domain/goal/src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/goal/service/ActionPlanServiceTest.kt [hunk 1]
```diff
@@ -20,33 +26,103 @@ class ActionPlanServiceTest {
...
-  @Test
-  fun `should retrieve action plan by prison number`() {
-    // Given
-    val actionPlan = aValidActionPlan()
-    val prisonNumber = aValidPrisonNumber()
-    given(persistenceAdapter.getActionPlan(any())).willReturn(actionPlan)
+  @Nested
+  inner class CreateActionPlan {
+    @Test
+    fun `should create action plan given prisoner does not already have an action plan`() {
```
Comment (mattwills-moj-digital) [inline domain/goal/src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/goal/service/ActionPlanServiceTest.kt:32]: 🙌

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/ActionPlanEntityMapper.kt [hunk 2]
```diff
@@ -13,7 +14,8 @@ import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.goal.ActionPl
 )
 interface ActionPlanEntityMapper {
   @ExcludeJpaManagedFields
-  fun fromDomainToEntity(actionPlan: ActionPlan): ActionPlanEntity
+  @GenerateNewReference
+  fun fromDomainDtoToEntity(createActionPlanDto: CreateActionPlanDto): ActionPlanEntity
```
Comment (mattwills-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/ActionPlanEntityMapper.kt:18]: How about just `fromDtoToEntity`? DomainDto reads a bit weird to me and I thought it was a typo at first :)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/MappingExclusions.kt [hunk 3]
```diff
@@ -16,3 +16,6 @@ annotation class ExcludeJpaManagedFieldsIncludingDisplayNameFields
 
 @Mapping(target = "reference", ignore = true)
 annotation class ExcludeReferenceField
+
+@Mapping(target = "reference", expression = "java(UUID.randomUUID())")
+annotation class GenerateNewReference
```
Comment (mattwills-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/MappingExclusions.kt:21]: Nice 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/ActionPlanEntityMapper.kt (hunk 2 above)
Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/ActionPlanEntityMapper.kt:18]: I could do, but the DTO classes live in the domain module
And I've have to rename it in loads of places!

I can do if you'd rather though ...

File: domain/goal/src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/goal/service/ActionPlanServiceTest.kt [hunk 4]
```diff
@@ -20,33 +26,103 @@ class ActionPlanServiceTest {
...
+
+      // Then
+      assertThat(retrievedActionPlan.prisonNumber).isEqualTo(prisonNumber)
+      assertThat(retrievedActionPlan.goals).isEmpty()
+      verify(persistenceAdapter).getActionPlan(prisonNumber)
+    }
+  }
+
+  @Nested
+  inner class GetActionPlanSummaries {
```
Comment (mattwills-moj-digital) [inline domain/goal/src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/goal/service/ActionPlanServiceTest.kt:108]: Thanks for completing my work! 🙄

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/MappingExclusions.kt (hunk 3 above)
Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/MappingExclusions.kt:21]: I thought you'd like that 👍 😁

**Review by mattwills-moj-digital — DISMISSED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/ActionPlanEntityMapper.kt (hunk 2 above)
Comment (mattwills-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/ActionPlanEntityMapper.kt:18]: Yeah it's ok. I might change it separately 😉 😂

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/MappingExclusions.kt (hunk 3 above)
Comment (mattwills-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/MappingExclusions.kt:21]: I do. I did consider the filename, but it's fine!

**Review by mattwills-moj-digital — APPROVED** [review]
LGTM

**Pull request merged**